homeassistant/components/mqtt/scene.py

```diff
@@ -0,0 +1,145 @@
+"""Support for MQTT scenes."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import mqtt, scene
+from homeassistant.components.scene import Scene
+from homeassistant.const import CONF_ICON, CONF_NAME, CONF_PAYLOAD_ON, CONF_UNIQUE_ID
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.reload import async_setup_reload_service
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import (
+    ATTR_DISCOVERY_HASH,
+    CONF_COMMAND_TOPIC,
+    CONF_QOS,
+    CONF_RETAIN,
+    DOMAIN,
+    PLATFORMS,
+    MqttAvailability,
+    MqttDiscoveryUpdate,
+)
+from .discovery import MQTT_DISCOVERY_NEW, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "MQTT Scene"
+DEFAULT_RETAIN = False
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_COMMAND_TOPIC): mqtt.valid_publish_topic,
+        vol.Optional(CONF_ICON): cv.icon,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PAYLOAD_ON): cv.string,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
+        vol.Optional(CONF_RETAIN, default=DEFAULT_RETAIN): cv.boolean,
+    }
+).extend(mqtt.MQTT_AVAILABILITY_SCHEMA.schema)
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType, config: ConfigType, async_add_entities, discovery_info=None
+):
+    """Set up MQTT scene through configuration.yaml."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    await _async_setup_entity(hass, config, async_add_entities, discovery_info)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up MQTT scene dynamically through MQTT discovery."""
+
+    async def async_discover(discovery_payload):
+        """Discover and add a MQTT scene."""
+        discovery_data = discovery_payload.discovery_data
+        try:
+            config = PLATFORM_SCHEMA(discovery_payload)
+            await _async_setup_entity(
+                hass, config, async_add_entities, config_entry, discovery_data
+            )
+        except Exception:
+            clear_discovery_hash(hass, discovery_data[ATTR_DISCOVERY_HASH])
+            raise
+
+    async_dispatcher_connect(
+        hass, MQTT_DISCOVERY_NEW.format(scene.DOMAIN, "mqtt"), async_discover
+    )
+
+
+async def _async_setup_entity(
+    hass, config, async_add_entities, config_entry=None, discovery_data=None
+):
+    """Set up the MQTT scene."""
+    async_add_entities([MqttScene(hass, config, config_entry, discovery_data)])
+
+
+class MqttScene(
+    MqttAvailability,
+    MqttDiscoveryUpdate,
+    Scene,
+):
+    """Representation of a scene that can be activated using MQTT."""
+
+    def __init__(self, hass, config, config_entry, discovery_data):
```

Comment:
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.

Reply:
I think this pattern is inherited from other MQTT component where `hass` is needed to setup templates. In this case there is no template support and it could be removed.

Reply:
Done